File: tourism-be/src/Modules/Tours/Explorer.Tours.Core/Domain/Tours/Tour.cs
using Explorer.BuildingBlocks.Core.Domain;
using Microsoft.IdentityModel.Tokens;

namespace Explorer.Tours.Core.Domain.Tours
{

    public class Tour : Entity
    {
        public string Name { get; private set; }
        public string Description { get; init; }
        public TourDifficulty Difficulty { get; init; }
        public List<string> Tags { get; init; }
        public TourStatus Status { get; private set; }
        public double Price { get; init; }
        public int AuthorId { get; init; }
        public int[] Equipment { get; init; }
        public double DistanceInKm { get; init; }
        public DateTime? ArchivedDate { get; private set; }
        public DateTime? PublishedDate { get; private set; }
        public List<TourDuration> Durations { get; private set; }
        public List<TourKeyPoint> KeyPoints { get; private set; }
        public Uri? Image { get; private set; }

        public Tour(string name, string description, TourDifficulty difficulty, List<string> tags, TourStatus status, double price, int authorId, int[] equipment, double distanceInKm, DateTime? archivedDate, DateTime? publishedDate, List<TourDuration> durations, Uri? image = null)
        {
            Name = name;
            Description = description;
            Difficulty = difficulty;
            Tags = tags;
            Status = status;
            Price = price;
            AuthorId = authorId;
            Equipment = equipment;
            DistanceInKm = distanceInKm;
            ArchivedDate = archivedDate;
            PublishedDate = publishedDate;
            Durations = durations;
            KeyPoints = new List<TourKeyPoint>();
            if (Status == TourStatus.TouristMade)
            {
                TouristTourValidation();
            }
            Image = image ?? new Uri("https://www.flimslaax.com/fileadmin/Daten/0Flims_Laax_Bilder/3-Outdoor/3-2-Wandern/3-2-1-Wanderwege/flims_laax_falera_wanderwege2.jpg");
        }

        private void Validate()
        {
            if (string.IsNullOrWhiteSpace(Name)) throw new ArgumentException("Invalid Name");
            if (string.IsNullOrWhiteSpace(Description)) throw new ArgumentException("Invalid Description");
            if (Price < 0) throw new ArgumentException("Invalid Price");
            if (Tags.IsNullOrEmpty()) throw new ArgumentException("Not enough Tags");
            if (KeyPoints.Count < 2) throw new ArgumentException("Not enough Key Points");
            if (Durations.IsNullOrEmpty()) throw new ArgumentException("Not enough Durations");
            if (Status == TourStatus.Published) throw new ArgumentException("Tour is already published");
        }

        private void TouristTourValidation()
        {
            if (string.IsNullOrWhiteSpace(Name)) throw new ArgumentException("Invalid Name");
            if (Durations.Count == 0) throw new ArgumentException("Not enough Durations");
            if (Status != TourStatus.TouristMade) throw new ArgumentException("Tourist didn't make this tour");
        }

        public void Publish(int userId)
        {
            Validate();
            IsAuthor(userId);

            PublishedDate = DateTime.UtcNow;
            Status = TourStatus.Published;
        }

        public void Archive(int userId)
        {
            if (Status != TourStatus.Published) throw new ArgumentException("Tour must be published in order to be archived");
            IsAuthor(userId);

            ArchivedDate = DateTime.UtcNow;
            Status = TourStatus.Archived;
        }

        private void IsAuthor(int userId)
        {
            if (AuthorId != userId) throw new UnauthorizedAccessException("User is not the author of the tour");
        }

        public void UpdateTouristTour(Tour tour)
        {
            Name = tour.Name;
            KeyPoints = tour.KeyPoints;
            Durations = tour.Durations;
        }
    }

    public enum TourStatus
    {
        Draft,
        Published,
        Archived,
        TouristMade
    }

    public enum TourDifficulty
    {
        Beginner,
        Intermediate,
        Advanced,
        Pro
    }
}

File: golang-be/Tours/model/Tour.go
package model

import (
	"database/sql/driver"
	"errors"
	"fmt"
	"strings"
	"time"
	//"gorm.io/gorm"
)

type ArrayString []string

// Value converts ArrayString to a PostgreSQL-compatible representation
func (a ArrayString) Value() (driver.Value, error) {
	if len(a) == 0 {
		return "{}", nil
	}

	// Convert array elements to a string slice
	strSlice := make([]string, len(a))
	for i, v := range a {
		strSlice[i] = fmt.Sprintf("%s", v)
	}

	// Join elements with commas and enclose in curly braces
	result := fmt.Sprintf("{%s}", strings.Join(strSlice, ","))

	return result, nil
}

// Scan converts a PostgreSQL array string representation to ArrayString
func (a *ArrayString) Scan(value interface{}) error {
	if value == nil {
		*a = nil
		return nil
	}

	// Convert value to string
	str, ok := value.(string)
	if !ok {
		return fmt.Errorf("unsupported type for ArrayString: %T", value)
	}

	// Remove curly braces from the string
	str = strings.TrimPrefix(str, "{")
	str = strings.TrimSuffix(str, "}")

	// Split string into individual tags
	tags := strings.Split(str, ",")

	// Trim whitespaces from each tag
	for i, tag := range tags {
		tags[i] = strings.TrimSpace(tag)
	}

	*a = ArrayString(tags)
	return nil
}

type Tour struct {
	Id            int            `bson:"_id,omitempty"` // If you want to use a custom ID, use "_id" field name
	Name          string         `bson:"name"`
	Description   string         `bson:"description"`
	Price         float64        `bson:"price"`
	Difficulty    TourDifficulty `bson:"difficulty"`
	Tags          []string       `bson:"tags"`
	Equipment     []int32        `bson:"equipment"`
	Status        TourStatus     `bson:"status"`
	AuthorID      int            `bson:"author_id"`
	DistanceInKm  float64        `bson:"distance_in_km"`
	ArchivedDate  time.Time      `bson:"archived_date"`
	PublishedDate time.Time      `bson:"published_date"`
	Image         string         `bson:"image"`
	KeyPoints     []TourKeyPoint `bson:"key_points"`
	Durations     []TourDuration `bson:"durations"`
}

func NewTour(name string, description string, price float64, difficulty TourDifficulty,
	tags ArrayString, status TourStatus, authorId int, equipment []int, distanceInKm float64,
	archiveDate, publishDate time.Time, image string, keyPoints []TourKeyPoint, durations []TourDuration) *Tour {
	tour := &Tour{
		Name:          name,
		Description:   description,
		Price:         price,
		Difficulty:    difficulty,
		Tags:          tags,
		Status:        status,
		AuthorID:      authorId,
		DistanceInKm:  distanceInKm,
		ArchivedDate:  archiveDate,
		PublishedDate: publishDate,
		Image:         image,
		KeyPoints:     keyPoints,
		Durations:     durations,
	}

	return tour
}

func (tour *Tour) validate() error {
	if tour.Name == "" {
		return errors.New("invalid name")
	}
	if tour.Description == "" {
		return errors.New("invalid Description")
	}
	if tour.Price < 0 {
		return errors.New("invalid Price")
	}
	if len(tour.Tags) == 0 {
		return errors.New("not enough Tags")
	}
	if len(tour.KeyPoints) < 2 {
		return errors.New("not enough Key Points")
	}

	if tour.Status == Published {
		return errors.New("tour is already published")
	}

	return nil
}

func (tour *Tour) Publish(authorId int) error {
	if tour.AuthorID != authorId {
		return errors.New("author id doesn't match tour")
	}
	tour.Status = Published
	return nil
}

func (tour *Tour) Archive(authorId int) error {
	if tour.AuthorID != authorId {
		return errors.New("author id doesn't match tour")
	}
	tour.Status = Archived
	return nil
}

type TourStatus int

const (
	Draft TourStatus = iota
	Published
	Archived
	TouristMade
)

type TourDifficulty int

const (
	Beginner TourDifficulty = iota
	Intermediate
	Advanced
	Pro
)
